synchronization/futex.rs

```diff
@@ -0,0 +1,524 @@
+                Ok(_) => {
+                    // we need to grab the lock.
+                    return self.inner_lock(mutex);
+                }
+                Err(NixError::Sys(errno)) => {
+                    if unlikely(errno == Errno::EAGAIN) {
+                        return self.inner_lock(mutex);
+                    }
+
+                    if unlikely(errno == Errno::EINTR) {
+                        continue;
+                    }
+
+                    panic!("Unexpected failure while waiting on futex: FUTEX_WAIT({:p}, {}, None) failed due to {}", &self, start, errno);
+                }
+                _ => {
+                    unreachable!("Should never get this error");
+                }
+            }
+        }
+    }
+
+    pub fn wake(&self, _mutex: &FutexMutex<Futex>, number: WakeNumber) -> i32 {
+        self.0.fetch_add(1, Ordering::SeqCst);
+
+        self.0.wake(number).expect("Failed to wake CV")
+    }
+}
+
+pub type Mutex<T> = lock_api::Mutex<FutexMutex<Futex>, T>;
+pub type MutexGuard<'a, T> = lock_api::MutexGuard<'a, FutexMutex<Futex>, T>;
+
+/// Wrapper around lower level synchronization primitives allowing
+/// `std::sync::CondVar`-like usage which can be sent through shared
+/// memory and work across multiple properties.
+pub struct CondVar<T> {
+    condition: FutexCondition,
+    mutex: Mutex<T>,
+}
+
+impl<T> CondVar<T> {
+    /// Returns a new condition variable that works with the mutex passed in.
+    pub fn new(mutex: Mutex<T>) -> CondVar<T> {
+        CondVar {
+            condition: FutexCondition(Futex(AtomicI32::new(0))),
+            mutex,
+        }
+    }
+
+    /// Wake up one task waiting on this condition variable.
+    ///
+    /// Returns `true` if a thread was woken up.
+    ///
+    /// To wake up all threads see `notify_all`.
+    pub fn notify_one(&self) -> bool {
+        let woken = unsafe { self.condition.wake(self.mutex.raw(), WakeNumber::N(1)) };
+
+        if woken == 1 {
+            true
+        } else if woken == 0 {
+            false
+        } else {
+            panic!("Woke more than one waiter in notify_one");
+        }
+    }
+
+    /// Wakes up all tasks waiting on this condition variable.
+    ///
+    /// Returns the number of threads it woke up.
+    ///
+    /// To wake up only one thread, see `notify_one`.
+    pub fn notify_all(&self) -> usize {
+        unsafe { self.condition.wake(self.mutex.raw(), WakeNumber::All) as usize }
+    }
+
+    /// Blocks the current task until the condition variable is notified.
+    ///
+    /// Unlocks the mutex passed in to the condition variable, and then sleeps.
+    /// When this function returns, the lock has been reacquired.
+    pub fn wait(&self) {
+        unsafe {
+            self.condition.wait(self.mutex.raw());
+        }
+    }
+
+    /// Returns the mutex that should be locked to use this condition variable.
+    pub fn mutex(&self) -> &Mutex<T> {
+        return &self.mutex;
+    }
```

Comment:
Should the internal mutex be publicly visible (as opposed to passing lock/unlock through the condition variable wrapper itself)?